### PR TITLE
saul: add PWM config for use in SAUL

### PR DESCRIPTION
**Pull request by kYc0o**
I noticed that the PWM configuration was not available for SAUL.

This PR adds the support and also a config file for the samr21-xpro board. The latter is optional since there are no other SAUL config files (e.g. ADC) besides the onboard LED and button, but it was useful while testing. Thus, it might be removed.

The driver can take four parameters: channel (implicit), frequency, resolution and duty cycle.

### CODE REVIEW

File: boards/samr21-xpro/include/pwm_params.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#define STEPS       (1000U)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief    PWM pin configuration
+ */
+static const  saul_pwm_params_t saul_pwm_params[] =
```
Comment (haukepetersen) [inline boards/samr21-xpro/include/pwm_params.h]: I don't think mapping all those PWM lines per default is a good idea, as the pin usage overlaps with other functions as e.g. I2C (PA 16) and the ADC (PA6 and PA7). Also there is IMHO no need to export all 5 PWM pins per default in the first place. I would suggest to only export pins PA18 and PA19, as these per default do not overlap with anything.

File: drivers/include/saul/periph.h [hunk 2]
```diff
@@ -53,6 +57,20 @@ typedef struct {
...
+ * @brief   Direct mapped PWM configuration values
+ */
+typedef struct {
+    const char *name;       /**< name of the device connected to this pin */
+    pwm_t line;             /**< PWM line to initialize and expose */
+    pwm_mode_t mode;        /**< PWM mode */
+    uint32_t freq;          /**< PWM frequency */
+    uint16_t res;           /**< PWM resolution */
+    uint8_t channel;        /**< PWM channel */
+} saul_pwm_params_t;
```
Comment (haukepetersen) [inline drivers/include/saul/periph.h]: This struct contains a lot of redundant information. The configuration should reflect the structure of the PWM devices -> n devices with x channels each.

File: sys/include/phydat.h [hunk 3]
```diff
@@ -107,7 +107,8 @@ enum {
     UNIT_PPM,       /**< part per million */
     /* aggregate values */
     UNIT_TIME,      /**< the three dimensions contain sec, min, and hours */
-    UNIT_DATE       /**< the 3 dimensions contain days, months and years */
+    UNIT_DATE,      /**< the 3 dimensions contain days, months and years */
+    UNIT_PWM        /**< the 3 dimensions contain duty cycle, frequency and resolution */
```
Comment (haukepetersen) [inline sys/include/phydat.h]: this does IMHO not make any sense. PWM is not a physical value in itself, so I don't see it as part of this list. What we do control when using a PWM is actually the duty cycle, so why not use `UNIT_PERMILL` as unit for the SAUL pwm driver?

File: sys/auto_init/saul/auto_init_pwm.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * @brief   Define the number of configured sensors
+ */
+#define SAUL_PWM_NUMOF    (sizeof(saul_pwm_params)/sizeof(saul_pwm_params[0]))
+
+/**
+ * @brief   Allocate memory for the device descriptors
+ */
+static const saul_pwm_params_t *saul_pwms[SAUL_PWM_NUMOF];
```
Comment (haukepetersen) [inline sys/auto_init/saul/auto_init_pwm.c]: no need for this buffer, we could reference the parameters defined by the board directly (see #7586).

File: drivers/saul/pwm_saul.c [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    if (value->val[2] > 0) {
+        res = value->val[2];
+    }
+    else {
+        res = params->res;
+    }
+
+    pwm_init(params->line, params->mode, freq, res);
```
Comment (haukepetersen) [inline drivers/saul/pwm_saul.c]: as this is a set functin (and the duty cycle being the set value), I don't see the init function being part of this. 

If the rationale here is to actually be able to adapt the PWM parameters using the SAUL interface, I think we need to re-think this approach and figure out something more elegant -> we might even have to re-visit the peripheral PWM interface, as it is in the current form not very suitable for on-the-fly adaption of the PWM parameters...

**Comment by haukepetersen** [comment]
Thinking about this PR some more, it seems not to work well with the general SAUL concept of offering 'endpoints' tied to physical values. As PWM is a generic output signal that is only used to drive certain physical attributes, it is conceptually not right to export PWM lines via SAUL directly. The 'correct' way would be to provide different drivers, e.g. one for RGB LEDs, one for DC motors, etc. This way the PWM initialization would also be (statically) tied to the actual hardware that is driven by the PWM...

**Comment by kYc0o** [comment]
Sorry for not addressing the comments, I'll try asap.

>Thinking about this PR some more, it seems not to work well with the general SAUL concept of offering 'endpoints' tied to physical values. As PWM is a generic output signal that is only used to drive certain physical attributes, it is conceptually not right to export PWM lines via SAUL directly. The 'correct' way would be to provide different drivers, e.g. one for RGB LEDs, one for DC motors, etc. This way the PWM initialization would also be (statically) tied to the actual hardware that is driven by the PWM...

Though I agree with you here, it can be discussed... there's saul drivers for adc and gpio, which can also represent a wide range of sensors and actuators, so I see that pwm might have its place too.

Actually, the goal was to control a small buzzer with this PR, so I have created a driver called buzzer, which I didn't push since for me pwm was more useful in itself.

**Comment by haukepetersen** [comment]
> Though I agree with you here, it can be discussed... there's saul drivers for adc and gpio, which can also represent a wide range of sensors and actuators, so I see that pwm might have its place too.

True. ADC is border line, but different compared to PWM as it does not depend on any one-time configuration data. GPIO is simply an example of bad naming -> this module implements actually a button and a switch driver...

**Comment by haukepetersen** [comment]
> Actually, the goal was to control a small buzzer with this PR, so I have created a driver called buzzer, which I didn't push since for me pwm was more useful in itself.

Then I would prefer to go with the buzzer driver and export that via SAUL!

**Comment by kYc0o** [comment]
Closing due to misunderstanding on how SAUL should work.

**Pull request closed without merging**